operator/pkg/psp/emm.go

```diff
@@ -0,0 +1,711 @@
+	if err != nil {
+		logger.Error(err, "Encountered an error while cleaning up defunct EMM failure annotations")
+		return result.Error(err)
+	}
+	if didUpdate {
+		logger.Info("Cleaned up defunct EMM failure annotations")
+		return result.RequeueSoon(2)
+	}
+
+	// Do not perform EMM operations while the datacenter is initializing
+	if !provider.IsInitialized() {
+		logger.Info("Skipping EMM check as the cluster is not yet initialized")
+		return result.Continue()
+	}
+
+	// Find tainted nodes
+	//
+	// We may have some tainted nodes where we had previously failed the EMM
+	// operation, so be sure to filter those out.
+	plannedDownNodeNameSet, err := provider.getPlannedDownTimeNodeNameSet()
+	if err != nil {
+		return result.Error(err)
+	}
+	evacuateDataNodeNameSet, err := provider.getEvacuateAllDataNodeNameSet()
+	if err != nil {
+		return result.Error(err)
+	}
+
+	// Fail any evacuate data EMM operations if the datacenter is stopped
+	//
+	// Cassandra must be up and running to rebuild cassandra nodes. Since 
+	// evacuating may entail deleting PVCs, we need to fail these operations
+	// as we are unlikely to be able to carry them out successfully.
+	if provider.IsStopped() {
+		didUpdate := false
+		for nodeName, _ := range evacuateDataNodeNameSet {
+			podsUpdated, err := provider.failEMM(nodeName, GenericFailure)
+			if err != nil {
+				return result.Error(err)
+			}
+			logger.Info("Failed EMM operation for evacuate all data as the datacenter is currently stopped", "node", nodeName)
+			didUpdate = didUpdate || podsUpdated
+		}
+
+		if didUpdate {
+			return result.RequeueSoon(2)
+		}
+	}
+
+	// NOTE: There might be pods that aren't ready for a variety of reasons,
+	// however, with respect to data availability, we really only care if a
+	// pod representing a cassandra node that is _currently_ bootstrapped to 
+	// the is down. We do not care if a pod is down for a cassandra node
+	// that was never bootstrapped (for example, maybe we are in the middle of 
+	// scaling up), as such pods are not part of the cluster presently and 
+	// their being down has no impact on data availability. Also, simply
+	// looking at the pod state label is insufficient here. The pod might be
+	// brand new, but represents a cassandra node that is already part of the
+	// cluster.
+	racksWithDownPods := provider.getRacksWithNotReadyPodsBootstrapped()
+
+	// If we have multipe racks with down pods we will need to fail any
+	// EMM operation as cluster availability is already compromised.
+	if len(racksWithDownPods) > 1 {
+		allTaintedNameSet := utils.UnionStringSet(plannedDownNodeNameSet, evacuateDataNodeNameSet)
+		didUpdate := false
+		for nodeName, _ := range allTaintedNameSet {
+			didUpdatePods, err := provider.failEMM(nodeName, TooManyExistingFailures)
+			if err != nil {
+				return result.Error(err)
+			}
+			logger.Info("Failed EMM operation as availability is already compromised due to multiple racks having not ready pods", "node", nodeName)
```

Comment:
I think I'd prefer this at Warn, if this logger has that.  Otherwise error would be best.  I would not want to be a user an have this sort of thing pass by as Info.

Reply:
We only have Info and Error, and Error needs an actual non nil err to log with... I think what's in the PR already is basically a warn

Reply:
We could use fmt.Errorf(msg) to generate an error from a string:
 
https://golang.org/pkg/fmt/#Errorf 

I am a bit uncomfortable about recording EMM failures as Info, because we are dealing with things that can lead to availability and data failure scenarios.

Reply:
This case only comes about when their are pods down in multiple racks already and a user sits down, says "frak it", and tries to put a k8s node in EMM anyway. I'd expect there to already be errors logged concerning the pods that are not ready. It is not an error for EMM to fail when availability is compromised. It _is_ an error for availability to be compromised for which I'd expect we already have an error log.